// eslint-plugin-testing-library/rules/prefer-explicit-assert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
var experimental_utils_1 = require("@typescript-eslint/experimental-utils");
var utils_1 = require("../utils");
var node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'prefer-explicit-assert';
var ALL_GET_BY_QUERIES = utils_1.ALL_QUERIES_METHODS.map(function (queryMethod) { return "get" + queryMethod; });
var isValidQuery = function (node, customQueryNames) {
    return ALL_GET_BY_QUERIES.includes(node.name) ||
        customQueryNames.includes(node.name);
};
var isAtTopLevel = function (node) {
    return node.parent.parent.type === 'ExpressionStatement';
};
exports.default = experimental_utils_1.ESLintUtils.RuleCreator(utils_1.getDocsUrl)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Suggest using explicit assertions rather than just `getBy*` queries',
            category: 'Best Practices',
            recommended: false,
        },
        messages: {
            preferExplicitAssert: 'Wrap stand-alone `getBy*` query with `expect` function for better explicit assertion',
            preferExplicitAssertAssertion: '`getBy*` queries must be asserted with `{{assertion}}`',
        },
        fixable: null,
        schema: [
            {
                type: 'object',
                additionalProperties: false,
                properties: {
                    assertion: {
                        type: 'string',
                    },
                    customQueryNames: {
                        type: 'array',
                    },
                },
            },
        ],
    },
    defaultOptions: [
        {
            customQueryNames: [],
        },
    ],
    create: function (context, _a) {
        var options = _a[0];
        var customQueryNames = options.customQueryNames, assertion = options.assertion;
        var getQueryCalls = [];
        return {
            'CallExpression Identifier': function (node) {
                if (isValidQuery(node, customQueryNames)) {
                    getQueryCalls.push(node);
                }
            },
            'Program:exit': function () {
                getQueryCalls.forEach(function (queryCall) {
                    var node = node_utils_1.isMemberExpression(queryCall.parent)
                        ? queryCall.parent
                        : queryCall;
                    if (isAtTopLevel(node)) {
                        context.report({
                            node: queryCall,
                            messageId: 'preferExplicitAssert',
                        });
                    }
                    else if (assertion) {
                        var expectation = node.parent.parent.parent;
                        if (expectation.type === 'MemberExpression' &&
                            expectation.property.type === 'Identifier' &&
                            expectation.property.name !== assertion) {
                            context.report({
                                node: expectation.property,
                                messageId: 'preferExplicitAssertAssertion',
                                data: {
                                    assertion: assertion,
                                },
                            });
                        }
                    }
                });
            },
        };
    },
});
